### DISCUSSION TITLE
quickjs vs quickjs-ng

### DISCUSSION THREAD

**Discussion by laoshaw** [quickjs-ng/quickjs#258] [Q&A] [answered]
I noticed both projects become active in the last 2-3 months, are they working together, or merging code from each other, or they're totally isolated now

**Comment by bnoordhuis** [comment]
Fabrice got active again after we started the fork. There's a fair amount of code sharing going on. His December release was like 50% my commits :-)

**Comment by saghul** [comment]
Maybe I shouldn't have emailed him after all :-P

I have been talking to Charlie to see what we can do going forward. Very firendly conversation, too.

**Comment by ABBAPOH** [comment]
Hello, what's the current status of both projects? Seems like NG has a bit more activity and I am wondering if something changed in recent months compared to status in the README

**Comment by chqrlie** [comment]
> Hello, what's the current status of both projects? Seems like NG has a bit more activity and I am wondering if something changed in recent months compared to status in the README

NG is more active at the moment, but I am going to backport all PRs to the upstream repository.  The goal is to converge both projects within 6 months.  I guess we should update the `README` file.

**Reply by ABBAPOH** [reply]
Hello again! Any news here? We use quicks in [Qbs](http://github.com/qbs/qbs) but it seems it is stalled again, should we migrate to QuickjsNG?

**Reply by saghul** [reply]
No updates from our side. We haven't discussed specifics yet.

**Reply by saghul** [reply]
At this point NG is better maintained, pick the one you need.

**Reply by p-bakker** [reply]
Maybe update the README.md with the latest info on how QuickJS and QuickJS-NG relate to each other and what the current view is of how these two projects move forward?

**Reply by p-bakker** [reply]
Given that the README.md now explicitly states quickjs-ng is a _fork_ of quickjs, without any mention of the desire on both sides to merge the efforts, can we assume both project will remain and follow their own path forward?

**Reply by bnoordhuis** [reply, marked as answer]
I think that's what it boils down to, yes. I suspect Charlie and Fabrice just don't have the time anymore to work on quickjs. Last bellard/master activity was 4 months ago and was minor stuff.

At this point quickjs-ng is so far ahead that talk of merging or converging also no longer makes sense. If we were to move forward together, we'd simply pick quickjs-ng as the new canonical source of truth.

**Reply by Sytten** [reply]
rquickjs and llrt switched to quickjs-ng, I think it should become the new "quickjs" v2 or something like that.

**Comment by papoteur-mga** [comment]
Hello,
I discover quickjs with yt-dlp which now needs a javascript backend.
They says "all versions [of quickjs-ng] are supported by yt-dlp; however, [upstream QuickJS](https://bellard.org/quickjs/) is recommended instead for performance reasons"
Do you know why they wrote that?
I'm a packager for Mageia and wonder which one to import. I'm successful to package yours for now.

**Reply by GalaxySnail** [reply]
See https://github.com/bellard/quickjs/issues/445 for details.

**Reply by bnoordhuis** [reply]
Also https://github.com/quickjs-ng/quickjs/issues/1002#issuecomment-3527428987 - in a nutshell: bellard/quickjs contains an optimization for pathological behavior in youtube's js blob that quickjs-ng/quickjs doesn't (yet) have because it comes with downsides of its own; sky high memory usage in particular.

I'll probably tackle it in a different way but no ETA because I don't have much free time at the moment.

**Comment by birdie-github** [comment]
Maybe I'm being stupid or slow today, but I haven't found the answers I'm looking for in this thread.

If you're here and see this message: _"QuickJS-NG is missing some optimizations making this very slow. Consider using upstream QuickJS instead"_ when using yt-dlp, you should use or install the original QuickJS implementation by Fabrice Bellard : https://github.com/bellard/quickjs (git clone, make -j`nproc`, sudo make install).

**Reply by Sytten** [reply]
This comment is very misleading, it is only faster because yt-dl uses a lot of small strings and quickjs has an optimization for that specific usecase called ropes. It is being ported over.
In general NG is faster.

**Reply by saghul** [reply]
FWIW, rope string _just_ landed in NG's master :-)

**Reply by avih** [reply]
Is it still unlikely for the two projects to merge in the forseeable future?

So basically users have to choose explicitly one or the other?

**Reply by saghul** [reply]
Fabrice seems to like to wok by himself, and we respect that. He was kind enough to make the project available to everyone under a permissive license, so we get the chance to improve it. We merge from each other these days.

**Reply by avih** [reply]
> We merge from each other these days.

That's good to know, and apparently it has worked reasonably well so far, but it can get harder over time.

It's a bit of a shame that they're maintained independently, and users have to pick...

No real complaints here, I get it, but still, not fun for users...

**Reply by birdie-github** [reply]
> This comment is very misleading, it is only faster because yt-dl uses a lot of small strings and quickjs has an optimization for that specific usecase called ropes. It is being ported over. In general NG is faster.

I installed QuickJS-NG specifically for yt-dlp and it's **unusable** for it. I waited 5 minutes for some YT JS code to run and I just gave up. QuickJS finished the same code in roughly **two seconds**.

Nothing was misleading about my comment. Many people install QuickJS/QuickJS-NG for youtube and nothing else and they couldn't care less than "in general the NG fork is faster". It has an enormous performance/usability gap for the scenario that's relevant.

**Reply by saghul** [reply]
Read above. That performance optimization was merged yesterday. There is no release with it out.

**Reply by birdie-github** [reply]
> Read above. That performance optimization was merged yesterday. There is no release with it out.

I do appreciate that the issue has finally been addressed (after how many months) but I simply grabbed the latest release, found out it was practically unusable and left a comment - exactly what most users would do. I didn't intend to badmouth your project, or anything, I just wanted to inform others of the current situation. I'm really sorry if you came to that conclusion. Cheers!

**Comment by saghul** [comment]
I'm going to close this since I don't think there is more to be added here.
